File: packages/node-common/src/storage.ts
import { S3Client, PutObjectCommand, GetObjectCommand } from "@aws-sdk/client-s3";
import { Readable } from "node:stream";

export function createStorage(): S3Client {
  return new S3Client({
    endpoint: process.env.OBJECT_STORE_ENDPOINT,
    region: process.env.OBJECT_STORE_REGION ?? "us-east-1",
    credentials: {
      accessKeyId: process.env.OBJECT_STORE_ACCESS_KEY!,
      secretAccessKey: process.env.OBJECT_STORE_SECRET_KEY!
    },
    forcePathStyle: true
  });
}

export const BUCKET = process.env.OBJECT_STORE_BUCKET ?? "audio-api";

export async function putObject(s3: S3Client, key: string, body: Buffer | Readable, contentType?: string): Promise<void> {
  await s3.send(new PutObjectCommand({ Bucket: BUCKET, Key: key, Body: body, ContentType: contentType }));
}

export async function getObjectStream(s3: S3Client, key: string): Promise<Readable> {
  const r = await s3.send(new GetObjectCommand({ Bucket: BUCKET, Key: key }));
  return r.Body as Readable;
}

File: packages/py-common/py_common/storage.py
import os
import boto3

BUCKET = os.environ.get("OBJECT_STORE_BUCKET", "audio-api")

def client():
    return boto3.client(
        "s3",
        endpoint_url=os.environ["OBJECT_STORE_ENDPOINT"],
        aws_access_key_id=os.environ["OBJECT_STORE_ACCESS_KEY"],
        aws_secret_access_key=os.environ["OBJECT_STORE_SECRET_KEY"],
        region_name=os.environ.get("OBJECT_STORE_REGION", "us-east-1"),
    )

def download_to_file(s3, key: str, dest_path: str) -> None:
    s3.download_file(BUCKET, key, dest_path)

def upload_bytes(s3, key: str, data: bytes, content_type: str = "application/json") -> None:
    s3.put_object(Bucket=BUCKET, Key=key, Body=data, ContentType=content_type)
